crates/oxc_transformer/src/es2017/async_to_generator.rs

```diff
@@ -208,24 +397,306 @@ impl<'a, 'ctx> AsyncToGenerator<'a, 'ctx> {
+                params,
+                ctx.ast.alloc(body),
+                generator_function_id,
+                ctx,
+            ));
+        }
+
+        let wrapper_scope_id = ctx.create_child_scope(ctx.current_scope_id(), ScopeFlags::Function);
+        // The generator function will move to inside wrapper, so we need
+        // to change the parent scope of the generator function to the wrapper function.
+        ctx.scopes_mut().change_parent_id(generator_function_id, Some(wrapper_scope_id));
+
+        let bound_ident = Self::create_bound_identifier(None, wrapper_scope_id, ctx);
+
+        let caller_function = {
+            let scope_id = ctx.create_child_scope(wrapper_scope_id, ScopeFlags::Function);
+            let params = Self::create_placeholder_params(&params, scope_id, ctx);
+            let statements = ctx.ast.vec1(Self::create_apply_call_statement(&bound_ident, ctx));
+            let body = ctx.ast.alloc_function_body(SPAN, ctx.ast.vec(), statements);
+            let id = Self::infer_function_id_from_variable_declarator(wrapper_scope_id, ctx);
+            let function = Self::create_function(id, params, body, scope_id, ctx);
+            let argument = Some(ctx.ast.expression_from_function(function));
+            ctx.ast.statement_return(SPAN, argument)
+        };
+
+        // Wrapper function
+        {
+            let statement = self.create_async_to_generator_declaration(
+                &bound_ident,
+                params,
+                ctx.ast.alloc(body),
+                generator_function_id,
+                ctx,
+            );
+            let mut statements = ctx.ast.vec_with_capacity(2);
+            statements.push(statement);
+            statements.push(caller_function);
+            let body = ctx.ast.alloc_function_body(SPAN, ctx.ast.vec(), statements);
+            let params = Self::create_empty_params(ctx);
+            let wrapper_function = Self::create_function(None, params, body, wrapper_scope_id, ctx);
+            // Construct the IIFE
+            let callee = ctx.ast.expression_from_function(wrapper_function);
+            Some(ctx.ast.expression_call(SPAN, callee, NONE, ctx.ast.vec(), false))
+        }
+    }
+
+    /// Infers the function id from [`Ancestor::VariableDeclaratorInit`].
+    #[inline]
+    fn infer_function_id_from_variable_declarator(
+        scope_id: ScopeId,
+        ctx: &mut TraverseCtx<'a>,
+    ) -> Option<BindingIdentifier<'a>> {
+        let Ancestor::VariableDeclaratorInit(declarator) = ctx.parent() else {
+            return None;
+        };
+        let Some(id) = declarator.id().get_binding_identifier() else { unreachable!() };
+        Some(
+            ctx.generate_binding(id.name.clone(), scope_id, SymbolFlags::FunctionScopedVariable)
+                .create_binding_identifier(ctx),
+        )
+    }
+
+    /// Creates a [`Function`] with the specified params, body and scope_id.
+    #[inline]
```

Comment:
`#[inline]` seems sensible here. This function just constructs a type, and inlining it makes it more likely that compiler will see it can construct the type directly in arena, rather than constructing on stack and then moving to the arena.